tremor-script/src/interpreter.rs

```diff
@@ -434,48 +439,100 @@ where
                     return error_need_obj(outer, segment, current.value_type(), &env.meta);
                 }
             }
+            // Next segment is an index: index into `current`, if it's an array
             Segment::Idx { idx, .. } => {
                 if let Some(a) = current.as_array() {
-                    let (start, end) = if let Some((start, end)) = subrange {
-                        // We check range on setting the subrange!
-                        (start, end)
+                    let range_to_consider = subrange.unwrap_or_else(|| a.as_slice());
+                    let idx = *idx;
+
+                    if let Some(c) = range_to_consider.get(idx) {
+                        current = c;
+                        subrange = None;
+                        continue;
                     } else {
-                        (0, a.len())
-                    };
-                    let idx = *idx as usize + start;
-                    if idx >= end {
-                        // We exceed the sub range
-                        let bad_idx = idx - start;
                         return error_array_out_of_bound(
                             outer,
                             segment,
                             &path,
-                            bad_idx..bad_idx,
+                            idx..idx,
+                            range_to_consider.len(),
                             &env.meta,
                         );
                     }
+                } else {
+                    return error_need_arr(outer, segment, current.value_type(), &env.meta);
+                }
+            }
+            // Next segment is an index range: index into `current`, if it's an array
+            Segment::Range {
+                range_start,
+                range_end,
+                ..
+            } => {
+                if let Some(a) = current.as_array() {
+                    let range_to_consider = subrange.unwrap_or_else(|| a.as_slice());
 
-                    if let Some(c) = a.get(idx) {
-                        current = c;
-                        subrange = None;
-                        continue;
-                    } else {
+                    // Helper closure for evaluating an expression to an `usize` index
```

Comment:
We have a `eval_to_string` function in https://github.com/wayfair-tremor/tremor-runtime/blob/master/tremor-script/src/interpreter/imut_expr.rs#L54 would it make sense to pull this out into `eval_to_usize`?

That might reduce the code in this function a bit which was always too large I think.

Reply:
That's a good idea.

I think it might help to do a small refactor to group the context of the interpreter into a struct, since there's a lot of code duplication due to dragging the `opts`, `env`, `event`, `state`, `meta`, and `local` in and out of each method there. Tiny refactor. Really, no biggie refactor. Oh well, https://github.com/m0n0chr0m3/tremor-runtime/commit/fc921c5fa151d4e2cc6756583f5a66c586907234, "306 additions and 451 deletions", reducing code duplication achieved.

Reply:
hah :D 
* for some definitions of tiny
looks good general I left one comment on it where I would be a bit concerned if we allocate and copy the whole struct for every time we switch between  with and without result.

Reply:
Resolved: `opts` are no longer in the context `struct`.